pallets/proof-of-stake/src/reward_info.rs

```diff
@@ -0,0 +1,174 @@
+		let activated_amount = self
+			.activated_amount
+			.checked_sub(liquidity_assets_removed)
+			.ok_or(Error::<T>::LiquidityCheckpointMathError)?;
+
+		let missing_at_checkpoint_new = self
+			.calculate_missing_at_checkpoint(current_time)
+			.ok_or(Error::<T>::LiquidityCheckpointMathError)?;
+
+		let activated_amount_new = self
+			.activated_amount
+			.checked_sub(liquidity_assets_removed)
+			.ok_or(Error::<T>::LiquidityCheckpointMathError)?;
+		let missing_at_checkpoint_after_burn = U256::from(activated_amount_new)
+			.checked_mul(missing_at_checkpoint_new)
+			.and_then(|v| v.checked_div(self.activated_amount.into()))
+			.ok_or(Error::<T>::LiquidityCheckpointMathError)?;
+
+		let user_current_rewards = self
+			.calculate_rewards(current_time, pool_ratio_current)
+			.ok_or(Error::<T>::CalculateRewardsMathError)?;
+		let total_available_rewards = user_current_rewards
+			.checked_add(self.rewards_not_yet_claimed)
+			.and_then(|v| v.checked_sub(self.rewards_already_claimed))
+			.ok_or(Error::<T>::LiquidityCheckpointMathError)?;
+
+		self.activated_amount = activated_amount;
+		self.pool_ratio_at_last_checkpoint = pool_ratio_current;
+		self.rewards_already_claimed = 0_u128;
+		self.missing_at_last_checkpoint = missing_at_checkpoint_after_burn;
+		self.rewards_not_yet_claimed = total_available_rewards;
+		self.last_checkpoint = current_time;
+		Ok(())
+	}
+
+	pub fn calculate_rewards(
+		&self,
+		current_time: u32,
+		pool_rewards_ratio_current: U256,
+	) -> Option<Balance> {
+		let pool_rewards_ratio_new =
+			pool_rewards_ratio_current.checked_sub(self.pool_ratio_at_last_checkpoint)?;
+
+		let user_rewards_base: U256 = U256::from(self.activated_amount)
+			.checked_mul(pool_rewards_ratio_new)?
+			.checked_div(U256::from(u128::MAX))?; // always fit into u128
+
+		let (cumulative_work, cummulative_work_max_possible) =
+			self.calculate_cumulative_work_max_ratio(current_time)?;
+
+		user_rewards_base
+			.checked_mul(cumulative_work)?
+			.checked_div(cummulative_work_max_possible)?
+			.try_into()
+			.ok()
+	}
+
+	fn calculate_cumulative_work_max_ratio(&self, current_time: u32) -> Option<(U256, U256)> {
+		let time_passed = current_time.checked_sub(self.last_checkpoint).unwrap();
+		// .ok_or(Error::<T>::PastTimeCalculation)?;
+		let mut cummulative_work = U256::from(0);
+		let mut cummulative_work_max_possible_for_ratio = U256::from(1);
+
+		if time_passed != 0 && self.activated_amount != 0 {
+			let liquidity_assets_amount_u256: U256 = self.activated_amount.into();
+
+			// whole formula: 	missing_at_last_checkpoint*106/6 - missing_at_last_checkpoint*106*precision/6/q_pow
+			// q_pow is multiplied by precision, thus there needs to be *precision in numenator as well
+
+			cummulative_work_max_possible_for_ratio =
+				liquidity_assets_amount_u256.checked_mul(U256::from(time_passed))?;
+
+			// whole formula: 	missing_at_last_checkpoint*Q*100/(Q*100-100) - missing_at_last_checkpoint*Q*100/(Q*100-100)*REWARDS_PRECISION/q_pow
```

Comment:
all commented code from this method should be removed I think

Reply:
these are actually valuable comments, we just should clean it up at some point to have html doc as we have it for bootstrap